### PR TITLE
fix(zetaclient): guard signer amount conversions

### LINKED ISSUE #4577: Add uint64 bounds check before Amount.Uint64() in chain signers
## Summary

Several chain signers call `params.Amount.Uint64()` without bounds checking. `cosmossdk.io/math.Uint.Uint64()` panics when the value exceeds max uint64. While this is not externally exploitable (source chain deposit events use u64 amounts), the panic bypasses the normal cancel/fallback flow and creates an infinite retry loop.

**HackenProof reports:** [ZCNode-244](https://dashboard.hackenproof.com/manager/companies/zetachain/zetachain-blockchain-software/reports/ZCNode-244), [ZCNode-246](https://dashboard.hackenproof.com/manager/companies/zetachain/zetachain-blockchain-software/reports/ZCNode-246)

## Affected Files

**Sui signer:**
- `zetaclient/chains/sui/signer/signer_tx.go` — `buildWithdrawAndCallTx` amount conversion

**Solana signer:**
- `zetaclient/chains/solana/signer/withdraw.go`
- `zetaclient/chains/solana/signer/withdraw_spl.go`
- `zetaclient/chains/solana/signer/execute.go`
- `zetaclient/chains/solana/signer/execute_spl.go`
- `zetaclient/chains/solana/signer/increment_nonce.go`

## Expected Behavior

If `params.Amount` exceeds `math.MaxUint64`, return an error instead of panicking. For Sui, this should trigger the cancel fallback (`broadcastCancelTx`). For Solana, this should trigger the nonce-increment fallback.

## Context

Found during HackenProof triage (ZCNode-244, ZCNode-246). Not externally exploitable since source chain deposit events cap amounts at u64, but worth fixing for robustness.

### PR DESCRIPTION
**Pull request by alan747271363-art**
Closes #4577

## Summary
- add checked uint64 amount conversion helpers for Sui and Solana signers
- return ErrInvalidWithdrawalAmount instead of panicking when outbound amounts exceed uint64
- keep cancel/fallback Solana transactions at amount 0 without converting the original oversized amount
- add regression coverage for the conversion boundary

## Tests
- docker run --rm -v "${repo}:/workspace" -v zeta-go-cache:/go/pkg/mod --workdir /workspace golang:1.23 sh -c "go test -tags pebbledb,ledger,test ./zetaclient/chains/solana/signer"
- docker run --rm -v "${repo}:/workspace" -v zeta-go-cache:/go/pkg/mod --workdir /workspace golang:1.23 sh -c "go test -tags pebbledb,ledger,test ./zetaclient/chains/sui/signer"

<h3>Greptile Summary</h3>

This PR replaces silent truncating `params.Amount.Uint64()` calls in the Solana and Sui chain signers with a checked `outboundAmountUint64` helper that returns `ErrInvalidWithdrawalAmount` when the big-integer amount exceeds the uint64 range, and keeps cancel/fallback transactions pinned at amount `0` without touching the potentially oversized value.

- New `outboundAmountUint64` helper uses `BigInt().BitLen() > 64` to gate conversion; the boundary logic is correct for all values 0–MaxUint64 and is covered by regression tests.
- All five Solana signer call-sites (`withdraw`, `withdraw_spl`, `execute`, `execute_spl`, `increment_nonce`) and the Sui `buildWithdrawAndCallTx` are updated consistently; the Sui regular-withdraw path already used `Amount.String()` and was unaffected.
- The helper is duplicated verbatim in both `solana/signer/amount.go` and `sui/signer/amount.go`; and the Sui `broadcastWithdrawalWithFallback` does not yet catch `ErrInvalidWithdrawalAmount` to fire the cancel tx when a withdrawAndCall amount is oversized.

<h3>Confidence Score: 4/5</h3>

Safe to merge — all five Solana call-sites and the Sui PTB path are correctly guarded, and the fix is strictly better than the previous silent truncation.

The Solana and Sui regular-withdrawal paths are handled cleanly. The two open items — duplicated helper code and the missing ErrInvalidWithdrawalAmount cancel arm in Sui's broadcastWithdrawalWithFallback — are follow-up improvements rather than regressions introduced by this PR.

zetaclient/chains/sui/signer/signer_tx.go deserves a second look around the broadcastWithdrawalWithFallback error switch to ensure the new error type is treated as a cancellable condition.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| zetaclient/chains/solana/signer/amount.go | New helper `outboundAmountUint64` with correct BitLen guard; logic is sound for 0–MaxUint64 range. |
| zetaclient/chains/solana/signer/amount_test.go | Tests cover MaxUint64 boundary and the first overflowing value (2^64); boundary conditions are correctly exercised. |
| zetaclient/chains/solana/signer/execute.go | Correctly defers amount conversion to the non-cancel path and propagates overflow error. |
| zetaclient/chains/solana/signer/increment_nonce.go | Guarded conversion applied correctly; cancel-tx path unconditionally uses amount=0 as intended. |
| zetaclient/chains/solana/signer/withdraw.go | Overflow guard added; cancel-tx continues to use amount=0 without touching the oversized value. |
| zetaclient/chains/sui/signer/amount.go | Byte-for-byte duplicate of solana/signer/amount.go; should be extracted to a shared package to avoid drift. |
| zetaclient/chains/sui/signer/amount_test.go | Byte-for-byte duplicate of the Solana test; same duplication concern. |
| zetaclient/chains/sui/signer/signer_tx.go | buildWithdrawAndCallTx now guards the amount; however ErrInvalidWithdrawalAmount is not caught in broadcastWithdrawalWithFallback to trigger a cancel — a stuck CCTX is possible if a withdrawAndCall ever carries an oversized amount. |

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[Outbound amount\nsdkmath.Uint] --> B{outboundAmountUint64\nBitLen > 64?}
    B -- Yes --> C[return ErrInvalidWithdrawalAmount]
    B -- No --> D[return uint64 value]

    subgraph Solana Signers
        E{cancelTx?}
        E -- true --> F[amount = 0\nskip conversion]
        E -- false --> G[call outboundAmountUint64]
        G --> H{error?}
        H -- yes --> I[return nil, err]
        H -- no --> J[proceed with tx]
    end

    subgraph Sui Signer
        K[buildWithdrawAndCallTx]
        K --> L[call outboundAmountUint64]
        L --> M{error?}
        M -- yes --> N[return error\nno cancel triggered]
        M -- no --> O[proceed with PTB]

        P[buildWithdrawTx]
        P --> Q[amount = params.Amount.String\nno conversion needed]
    end

    D --> E
    D --> K
    D --> P
```

<details open><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `zetaclient/chains/sui/signer/signer_tx.go`, line 278-290 ([link](https://github.com/zeta-chain/node/blob/7c3c676e01fcc756dde7fea967689eb70c47b5e1/zetaclient/chains/sui/signer/signer_tx.go#L278-L290)) 

   <a href="#">P2</a> **`ErrInvalidWithdrawalAmount` not handled — cancel tx never fires for oversized withdrawAndCall amounts**

   `broadcastWithdrawalWithFallback` cancels for `ErrInvalidPayload` and `ErrObjectOwnership`, but the new `ErrInvalidWithdrawalAmount` from `buildWithdrawAndCallTx` falls into the plain `case err != nil` branch and returns the error without cancelling. If a withdrawAndCall CCTX ever carries an amount > MaxUint64, the outbound will retry indefinitely and never self-resolve. Adding a `case errors.Is(err, cctypes.ErrInvalidWithdrawalAmount)` arm that invokes `s.broadcastCancelTx` would align this path with how other unrecoverable build errors are handled.

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: zetaclient/chains/sui/signer/signer_tx.go
   Line: 278-290

   Comment:
   **`ErrInvalidWithdrawalAmount` not handled — cancel tx never fires for oversized withdrawAndCall amounts**

   `broadcastWithdrawalWithFallback` cancels for `ErrInvalidPayload` and `ErrObjectOwnership`, but the new `ErrInvalidWithdrawalAmount` from `buildWithdrawAndCallTx` falls into the plain `case err != nil` branch and returns the error without cancelling. If a withdrawAndCall CCTX ever carries an amount > MaxUint64, the outbound will retry indefinitely and never self-resolve. Adding a `case errors.Is(err, cctypes.ErrInvalidWithdrawalAmount)` arm that invokes `s.broadcastCancelTx` would align this path with how other unrecoverable build errors are handled.

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
Fix the following 2 code review issues. Work through them one at a time, proposing concise fixes.

---

### Issue 1 of 2
zetaclient/chains/sui/signer/amount.go:10-14
**Duplicated helper — identical to `solana/signer/amount.go`**

`outboundAmountUint64` (and its test file) is a byte-for-byte copy of the Solana version. If the guard logic ever needs to change (e.g. tighter validation, a different error type), both copies must be updated in sync. Consider extracting this into a shared package — for example `zetaclient/chains/base` or a new `zetaclient/signerutils` — so both chain signers import the single canonical implementation.

### Issue 2 of 2
zetaclient/chains/sui/signer/signer_tx.go:278-290
**`ErrInvalidWithdrawalAmount` not handled — cancel tx never fires for oversized withdrawAndCall amounts**

`broadcastWithdrawalWithFallback` cancels for `ErrInvalidPayload` and `ErrObjectOwnership`, but the new `ErrInvalidWithdrawalAmount` from `buildWithdrawAndCallTx` falls into the plain `case err != nil` branch and returns the error without cancelling. If a withdrawAndCall CCTX ever carries an amount > MaxUint64, the outbound will retry indefinitely and never self-resolve. Adding a `case errors.Is(err, cctypes.ErrInvalidWithdrawalAmount)` arm that invokes `s.broadcastCancelTx` would align this path with how other unrecoverable build errors are handled.


`````

</details>

<sub>Reviews (1): Last reviewed commit: ["fix(zetaclient): guard signer amount con..."](https://github.com/zeta-chain/node/commit/7c3c676e01fcc756dde7fea967689eb70c47b5e1) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=31992039)</sub>

> Greptile also left **1 inline comment** on this PR.

### CODE REVIEW

File: zetaclient/chains/sui/signer/amount.go [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+	sdkmath "cosmossdk.io/math"
+
+	cctypes "github.com/zeta-chain/node/x/crosschain/types"
+)
+
+func outboundAmountUint64(amount sdkmath.Uint) (uint64, error) {
+	if amount.BigInt().BitLen() > 64 {
+		return 0, cosmoserrors.Wrap(cctypes.ErrInvalidWithdrawalAmount, "amount exceeds uint64 range")
+	}
+	return amount.Uint64(), nil
```
[the remark this answers is not in the document]
Comment (alan747271363-art) [reply zetaclient/chains/sui/signer/amount.go]: Fixed in `5acdb9d8`: the duplicated Sui/Solana helper has been extracted to `zetaclient/chains/base.OutboundAmountUint64`, with the shared tests moved to `base`. Both Sui and Solana signer callsites now use the shared helper, and the duplicate chain-local helper files were removed.

Local checks: `git diff --check` passes. I could not run `gofmt`/`go test` in this environment because the Go toolchain is not installed.
